Compatibility Verification

CyB3RC0nN0R opened the pull request:
This branch integrates the refactoring from `envoy-common` and verifies if the client version is bounded between the minimal and maximal version supported by the server.

The only currently supported version is `0.1-beta`.

Closes #39

DieGurke commented on src/main/java/envoy/server/data/GroupMessage.java:
```diff
@@ -24,6 +21,7 @@
 	@ElementCollection
 	private Map<Long, envoy.data.Message.MessageStatus> memberMessageStatus;
 
+	@Column(name = "last_status_change_date")
```
Isnt that inside the other pull request as well?

CyB3RC0nN0R replied:
Yes it is, because this branch is based on the previous one. For that reason this pull request is still a draft.

DieGurke replied:
ok

DieGurke commented on src/main/java/envoy/server/data/Message.java:
```diff
@@ -49,14 +50,14 @@
+	protected LocalDateTime creationDate;
 
-	@Temporal(TemporalType.TIMESTAMP)
-	protected Date receivedDate;
+	@Column(name = "received_date")
+	protected LocalDateTime receivedDate;
 
-	@Temporal(TemporalType.TIMESTAMP)
-	protected Date readDate;
+	@Column(name = "read_date")
+	protected LocalDateTime readDate;
```
Isnt that inside the other pull request as well?

DieGurke commented on src/main/java/envoy/server/data/User.java:
```diff
@@ -63,24 +62,19 @@
 
-	@Temporal(TemporalType.TIMESTAMP)
-	private Date lastSeen;
+	@Column(name = "last_seen")
```
Isnt that inside the other pull request as well?

CyB3RC0nN0R commented:
> 
> 
> I cant approve this before you explained to me why you rewrote basically the whole `LoginCredentialsProcessor` without making any real algorithm changes besides adding verification functionality!

There are three major reasons behind the "rewrite" of `LoginCredentialProcessor`:

- **Unnecessary functions:** the code contained several functions to handle user creation and retrieval, which were only called in one place, had unnecessarily long names and made it more difficult to understand whether a handshake has been rejected or not. These have been converted into an `if`-statement with two branches, which make the flow of the login process easier to understand, at least in my opinion. 

- **Efficiency:** the `toCommon` methods for both `User` and `Message` were called multiple times for the same objects, which I've replaced which a variable.

- **Uniform Logging:** although some logging was already present inside the class, several logger statements were unnecessary (e.g. _Sending Message[...]_, _Received LoginCredentials[...]_), because they are already present in `WriteProxy` and `ObjectMessageProcessor` respectively. However, I've missed some logger statements to track the login / registration process for a user and to understand how and why a handshake has been rejected. These are now added, which in conjunction with the aforementioned cleanup leads to an easily readable and understandable, redundance-free log.

delvh commented on src/main/java/envoy/server/processors/LoginCredentialProcessor.java:
```diff
@@ -29,126 +34,148 @@
+		// TODO: Verify compatibility
+		if (!verifyCompatibility(credentials.getClientVersion())) {
+			logger.info("The client has the wrong version.");
+			writeProxy.write(socketID, new HandshakeRejection("Wrong version"));
```
IIRC, the `HandshakeRejection` class offers default Strings as reasons.
It might be safer to replace the String with a default version in the original class.

- [ ] Additionally, one could insert a default reason of 'Username taken' there.

delvh commented on src/main/java/envoy/server/processors/LoginCredentialProcessor.java:
```diff
@@ -29,126 +34,148 @@
 
-	private static final Logger logger = EnvoyLog.getLogger(LoginCredentialProcessor.class);
+	private static final Logger		logger			= EnvoyLog.getLogger(LoginCredentialProcessor.class);
+	private static final Pattern	versionPattern	= Pattern.compile("(?<major>\\d).(?<minor>\\d)(?:-(?<suffix>\\w+))?");
```
- [ ] Regarding data encapsulation, it might be better to convert the verification compatibility processes into a static utility class.

DieGurke commented:
> 
> 
> > I cant approve this before you explained to me why you rewrote basically the whole `LoginCredentialsProcessor` without making any real algorithm changes besides adding verification functionality!
> 
> There are three major reasons behind the "rewrite" of `LoginCredentialProcessor`:
> 
>     * **Unnecessary functions:** the code contained several functions to handle user creation and retrieval, which were only called in one place, had unnecessarily long names and made it more difficult to understand whether a handshake has been rejected or not. These have been converted into an `if`-statement with two branches, which make the flow of the login process easier to understand, at least in my opinion.
> 
>     * **Efficiency:** the `toCommon` methods for both `User` and `Message` were called multiple times for the same objects, which I've replaced which a variable.
> 
>     * **Uniform Logging:** although some logging was already present inside the class, several logger statements were unnecessary (e.g. _Sending Message[...]_, _Received LoginCredentials[...]_), because they are already present in `WriteProxy` and `ObjectMessageProcessor` respectively. However, I've missed some logger statements to track the login / registration process for a user and to understand how and why a handshake has been rejected. These are now added, which in conjunction with the aforementioned cleanup leads to an easily readable and understandable, redundance-free log.

OK now im fine with this but i am not sure if a variable holding the common obj is good because i think maybe some variables in the server object change throughout the process of processing the login credentials. I hope you kept in mind, that this could happen between two toCommon calls so a variable would hold an old version of the server object (as interface object) and sending it. I am not sure if this happens here so it may be fine but i wanted to remind you of this incident, as it happens like this in other processors!

CyB3RC0nN0R commented:
> 
> 
> > > I cant approve this before you explained to me why you rewrote basically the whole `LoginCredentialsProcessor` without making any real algorithm changes besides adding verification functionality!
> > 
> > 
> > There are three major reasons behind the "rewrite" of `LoginCredentialProcessor`:
> > ```
> > * **Unnecessary functions:** the code contained several functions to handle user creation and retrieval, which were only called in one place, had unnecessarily long names and made it more difficult to understand whether a handshake has been rejected or not. These have been converted into an `if`-statement with two branches, which make the flow of the login process easier to understand, at least in my opinion.
> > 
> > * **Efficiency:** the `toCommon` methods for both `User` and `Message` were called multiple times for the same objects, which I've replaced which a variable.
> > 
> > * **Uniform Logging:** although some logging was already present inside the class, several logger statements were unnecessary (e.g. _Sending Message[...]_, _Received LoginCredentials[...]_), because they are already present in `WriteProxy` and `ObjectMessageProcessor` respectively. However, I've missed some logger statements to track the login / registration process for a user and to understand how and why a handshake has been rejected. These are now added, which in conjunction with the aforementioned cleanup leads to an easily readable and understandable, redundance-free log.
> > ```
> 
> OK now im fine with this but i am not sure if a variable holding the common obj is good because i think maybe some variables in the server object change throughout the process of processing the login credentials. I hope you kept in mind, that this could happen between two toCommon calls so a variable would hold an old version of the server object (as interface object) and sending it. I am not sure if this happens here so it may be fine but i wanted to remind you of this incident, as it happens like this in other processors!

I have considered this and can assure you that the described malfunction cannot occur because each common object is only used throughout one loop iteration, so no actual caching is performed.

delvh commented on src/main/java/enovy/server/util/VersionUtils.java:
```diff
@@ -0,0 +1,94 @@
+ * File: <strong>VersionUtils.java</strong><br>
+ * Created: <strong>23.06.2020</strong><br>
+ * 
+ * @author KSKE
```
```suggestion
 * @author Kai S. K. Engelbart
```
